### PR TITLE
inverse distance weighting interpolation for CICE

### LINKED ISSUE #833: bug: CICE get_expected_agreg_over_grid giving > 100%
:bug: 🧊 

User reported

### Describe the bug

1. Run cice with QCEFF with bounds on sea ice concentration 
2. What was the expected outcome? Run to completion, bounds respected
3. What actually happened?  CICE fwd op returning out of bounds causing QCEFF filter to die with "Largest ensemble member greater than upper bound"

### Error Message  
"Largest ensemble member greater than upper bound" 

### Which model(s) are you working with?
CESM CICE5(6?)

### Version of DART      
Which version of DART are you using? 
You can find the version using `git describe --tags`  
Will reproduce, believe it is v11

### Have you modified the DART code?   
No  

Work around is forcing the fwd op to be <100%

### Build information
Please describe:  
 1. Derecho
 2. unknown, will confirm if it makes a difference.

### PR DESCRIPTION
**Pull request by hkershaw-brown**
## Description:

Switched out bilinear interpolation for inverse distance weighting interpolation in CICE model_mod.f90
Power for IDW is fixed at 2.
The inverse distance weight keeps the interpolation result with the range of the 4 corners of the quad. 

### Fixes issue

fixes #833 

### Types of changes

- [x] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation update

### Documentation changes needed?

- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly.

### Tests
Please describe any tests you ran to verify your changes.

Ran the CICE egregious bounds case.
I know this is not sufficient to test the interpolation, but getting the ball rolling with this pull request. 

## Checklist for merging

- [x] Updated changelog entry
- [ ] Documentation updated
- [x] Update conf.py

## Checklist for release
- [ ] Merge into main
- [ ] Create release from the main branch with appropriate tag
- [ ] Delete feature-branch

## Testing Datasets

- [ ] Dataset needed for testing available upon request
- [ ] Dataset download instructions included
- [ ] No dataset needed

### CODE REVIEW

File: models/cice/model_mod.f90
Comment (jlaucar) [inline models/cice/model_mod.f90:1]: The get_dist function is able to handle any input longitude values correctly. There is no need to worry about whether they are overlapping at the 0 radian boundary. I think we can get rid of the checks and modifications of the longitudes. This will also remove the need for a Lon_in argument and the associated copy to a lon local variable.

Comment (hkershaw-brown) [reply models/cice/model_mod.f90:1]: removed the wrap around checks in 3b6a4323f98561cbb740f6d25e7f01c1771e7a53

**Comment by hkershaw-brown** [comment]
Don't merge as is (I can't review this since it I put the pull request in). Changes coming from @jlaucar

File: models/cice/model_mod.f90 [hunk 1]
```diff
@@ -1643,145 +1644,55 @@ end subroutine line_intercept
...
-function deter3(m)
- real(r8), intent(in) :: m(3, 3)
- real(r8)             :: deter3
+! Loop over the 4 corner points
+do i = 1, 4
+
+   distances(i) = get_dist(point, corner(i), no_vert=.true.)
 
-! Computes determinant of 3x3 matrix m
+   ! Handle division by zero (if the point coincides with a corner point)
```
[the remark this answers is not in the document]
Comment (Nuo-Chen) [reply models/cice/model_mod.f90:1681]:
```
    ! Check each corner for coincidence with the interpolation point
    do i = 1, 4
       distances(i) = get_dist(point, corner(i), no_vert=.true.)
       ! If point coincides with a corner, return the corresponding value immediately
       if (distances(i) < epsilon) then
           expected_obs = p(i)
           return
       endif
    end do
    
    ! Initialize sums
    sum_weights = 0.0_r8
    sum_weighted_p = 0.0_r8
    
    ! Calculate weights and sums if no coinciding corner was found
    do i = 1, 4
       weights(i) = 1.0_r8 / (distances(i)**power)
       sum_weights = sum_weights + weights(i)
       sum_weighted_p = sum_weighted_p + weights(i) * p(i)
    end do
```

I tried to impletment the return early option and it works with the 30 member ensemble test case.

File: models/cice/model_mod.f90 [hunk 2]
```diff
@@ -1643,145 +1644,60 @@ end subroutine line_intercept
...
+   distances(i) = get_dist(point, corner(i), no_vert=.true.)
 enddo
 
-end subroutine mat3x3
-
-!------------------------------------------------------------
+if(minval(distances) < epsilon_radians) then
+   ! To avoid any round off issues, if smallest distance is less than epsilon radians
+   ! just assign the value at the closest gridpoint to the interpolant
+   expected_obs = p(minloc(distances,1))
```
Comment (hkershaw-brown) [inline models/cice/model_mod.f90:1677]: @jlaucar not sure if you want a return here, or you want to have this case also go through the test-out-of-bounds

Comment (jlaucar) [reply models/cice/model_mod.f90:1677]: Let's go ahead and let it go through the out of bounds check. Obviously, it should never fail that, but this case is a boundary case that might not be that rare in some applications.

Comment (jlaucar) [reply models/cice/model_mod.f90:1677]: I have completed a suite of tests on the ensemble version in my branch on this algorithm. I never violated the bounds and always appropriately fit the quad corners. I think this should now be tested in a real CICE case.

Comment (hkershaw-brown) [reply models/cice/model_mod.f90:1677]: Awesome, I ran this lastest code with the original out-of-bounds CICE reproduce  from Nuo's , and it completed successfully (no out-of-bounds error)

**Review by jlaucar — APPROVED** [review]
I think this is ready to go, pending tests from Nuo. I tested in idealized settings to confirm the bounds and the general nature of the interpolation.

**Comment by Nuo-Chen** [comment]
The code also passed the 30 member CESM ensemle test on my end.

**Review by Nuo-Chen — APPROVED** [review]
The inverse distance weighting interpolation function fixed the out-of-bound issue created by the quadrilateral Interpolation for sea ice concentration.
